### ISSUE TITLE
Hook new_client can't move client to other workspace.

### ISSUE THREAD

**Issue by sbc64** [sminez/penrose#80]
I would like to begin with saying that the api that penrose exposes is great.  I'm quite excited to be building on top of it. I managed to replicate my dwm config almost immediately with penrose. The hooks are a great idea and I'm loving all the things I will be able to do.

### Describe the bug

I'm trying to use the `new_client()` Hook used in the example to push clients to the workspaces I want them to be in. 

```rust
struct MyClientHook {}
impl Hook for MyClientHook {
    fn new_client(&mut self, wm: &mut WindowManager, c: &mut Client) {
        wm.log(&format!("new client with WM_CLASS='{}'", c.wm_class()));
        wm.log(&format!("new client with WM_NAME='{}'", c.wm_name()));
        match c.wm_name().as_ref() {
            "calibre" => c.set_workspace(6),
            "Firefox Developer Edition" => {
                wm.log(&format!("Matched client name: {}", c.wm_name()));
                wm.log(&format!("Current Workspace: {}", c.workspace()));
                c.set_workspace(7);
                wm.log(&format!("Current Workspace: {}", c.workspace()));
                //wm.client_to_workspace(&Selector::Index(7));
            }
            _ => println!("something else!")
        }
    }

```

But the new client Firefox doesn't seem to go to the workspace 7 that I told it to go to. Should I use a different workspace function? `wm.client_to_workspace()` works in moving a sclient to a new workspace but it doesn't work when there is an already existing client in the workspace I opened Firefox since it moves the existing focused client to workspace.

### To Reproduce

1. Run dmenu_run with the keybinding: `M-p`
2. Select firefox-dev

### Expected behavior

I would expect Firefox Dev Edition to be moved to workspace number 7 but it stays on the workspace I'm currently in.

### Screenshots

https://vimeo.com/489017117

### Versions & OS Details

- OS: Linux
- Distribution NixOS
- OS Version 19.03
- Penrose Version `588c777`

### Additional context

Debug log:

```
16:51:12 [WARN] attempt to remove unknown client 44040196
16:51:13 [INFO] new client with WM_CLASS='simplescreenrecorder'
16:51:13 [INFO] new client with WM_NAME='SimpleScreenRecorder'
something else!
16:51:40 [INFO] new client with WM_CLASS='simplescreenrecorder'
16:51:40 [INFO] new client with WM_NAME='SimpleScreenRecorder'
something else!
16:51:45 [WARN] attempt to remove unknown client 44040211
16:51:45 [WARN] attempt to remove unknown client 44040211
16:51:52 [WARN] attempt to remove unknown client 54525956
16:51:52 [WARN] attempt to remove unknown client 56623105
16:51:52 [WARN] attempt to remove unknown client 54525954
16:51:52 [INFO] new client with WM_CLASS='Navigator'
16:51:52 [INFO] new client with WM_NAME='Firefox Developer Edition'
16:51:52 [INFO] Matched client name: Firefox Developer Edition
16:51:52 [INFO] Current Workspace: 1
16:51:52 [INFO] Current Workspace: 7
```

My [main.rs](https://github.com/sebohe/my-penrose/blob/develop/src/main.rs)

If you have a better way of having newly spawned clients to be moved to a default workspace I will gladly implement it. I will also start digging through penrose to try to understand it more. Thanks for reading!

**Comment by sminez** [comment]
Hey @sebohe, thank you very much for all of the detail on this issue! It looks like you are right about this not being possible without some rather over the top code at the moment (I think the _cleanest_ I can come up with so far is to cache the current focused client, focus the new one, move it using `client_to_workspace` and then reset the focus...). That said, it might be as simple as using the `workspace` property on the client instead of `wix` [here](https://github.com/sminez/penrose/blob/develop/src/core/manager.rs#L345) and [here](https://github.com/sminez/penrose/blob/develop/src/core/manager.rs#L353-L354)? I thought I'd made sure this was possible at one point but clearly not...I'll have a play around later and see if that does the trick :+1: 

(Note that this would _only_ work in the `new_client` hook but that shouldn't be an issue I don't think)

**Comment by sbc64** [comment]
Thanks for the pointers. :smiley:  I will see what I can cook up.

**Comment by sminez** [comment]
I haven't tested this yet, but #81 is what I'm thinking for a small change to enable this and a way of writing the hook itself to be usable for arbitrary clients.

**Comment by sbc64** [comment]
#81 does the trick for me. This is what I did in my `main.rs`:

```rust
config.hooks.push(ClientSpawnRules::new(vec![
  SpawnRule::WMName("Firefox Developer Edition" , 7),
]));
```

And it seems to work. I will add more clients... and see what happens and will report back if anything breaks or if it is all working.

**Comment by sbc64** [comment]
Closing the issue since the issue is resolved with PR #81. Thanks!
